### PR TITLE
Added FCHL19 representation. Not working for periodic systems so far.

### PR DESCRIPTION
**Pull request by FelixFaber**
I added FCH19 representation.

### CODE REVIEW

File: scripts/gen_FCHL19_rep.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/python3
```
Comment (Ryan-Rhys) [inline scripts/gen_FCHL19_rep.py]: If we're going to be adhering to PEP8

https://www.python.org/dev/peps/pep-0008/

I believe we'd require a module(script)-level docstring here.

Comment (Ryan-Rhys) [reply scripts/gen_FCHL19_rep.py]: Would be interesting to see what Pylint has to say

File: scripts/gen_FCHL19_rep.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+import argparse
+import os
+import sys
+
+import numpy as np
+from ase.io import read, write
```
Comment (Ryan-Rhys) [inline scripts/gen_FCHL19_rep.py]: @BingqingCheng Are we treating ase and asaplib as third-party libraries or project-level?

File: scripts/gen_FCHL19_rep.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+def repr_wrapper(frame, elements, is_periodic = False,
+                                        nRs2=24, nRs3=20,
+                                        nFourier=1, eta2=0.32, eta3=2.7,
+                                        zeta=np.pi, rcut=8.0, acut=8.0,
+                                        two_body_decay=1.8, three_body_decay=0.57,
+                                        three_body_weight=13.4):
+    '''
+   Periodic systems not implemented for FCHL19.
+
```
Comment (Ryan-Rhys) [inline scripts/gen_FCHL19_rep.py]: I believe Pylint should catch formatting-based things like the extra line here and the absence of a space between comma-separated function arguments. Would it be possible to include a docstring describing what the arguments are?

File: scripts/gen_FCHL19_rep.py (hunk 2 above)
Comment (BingqingCheng) [reply scripts/gen_FCHL19_rep.py]: What is the difference between the two?

File: scripts/gen_FCHL19_rep.py (hunk 3 above)
Comment (BingqingCheng) [reply scripts/gen_FCHL19_rep.py]: Felix, we then need to add instructions on how to install/import qml.

File: scripts/gen_FCHL19_rep.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        if output == 'matrix':
+            with open(foutput + ".desc", "ab") as f:
+                np.savetxt(f, frame.info[desc_name][None])
+            if peratom or nframes == 1:
+                with open(foutput + ".atomic-desc", "ab") as fatomic:
+                    np.savetxt(fatomic, rep)
+        elif output == 'xyz':
+            # output per-atom info
+            if peratom:
+                frame.new_array(desc_name, rep)
```
Comment (BingqingCheng) [inline scripts/gen_FCHL19_rep.py]: Does it output per atom representation (for each atomic environment)?

**Comment by BingqingCheng** [comment]
@FelixFaber , can you add an example on how to use it? You can edit the README.md file in the example/small_molecules.

**Comment by FelixFaber** [comment]
@BingqingCheng The atomic version should work. I have added an example of how to use it in the readme file. The QML code can be installed from https://www.qmlcode.org/installation.html. You will need to install the development branch.

**Comment by BingqingCheng** [comment]
Cool. I think we can merge.

File: scripts/gen_FCHL19_rep.py (hunk 2 above)
Comment (Ryan-Rhys) [reply scripts/gen_FCHL19_rep.py]: There'll be a difference in the import block in which they're included. Third-party libraries are in a separate block to project-level imports.

**Comment by Ryan-Rhys** [comment]
Still need the Pylint changes and the update README with instructions for qml.

File: scripts/gen_FCHL19_rep.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+import os
+import sys
+
+import numpy as np
+from ase.io import read, write
+from asaplib.io import str2bool
+from qml.representations import generate_fchl_acsf
+
+def repr_wrapper(frame, elements, is_periodic = False,
+                                        nRs2=24, nRs3=20,
```
Comment (Ryan-Rhys) [inline scripts/gen_FCHL19_rep.py]: A description of these parameters would be good. If the user is allowed to change them we should include a pointer to the parameter descriptions in the qml library. If the user isn't expected to change these values we probably shouldn't be giving them the option to do so.

File: scripts/gen_FCHL19_rep.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    
+def main(fxyz = False, fdict = False, prefix= False , output= False , peratom= False, 
+                    nRs2=24, nRs3=20, nFourier=1,
+                    eta2=0.32, eta3=2.7, zeta=np.pi, 
+                    rcut=8.0, acut=8.0,
+                    two_body_decay=1.8, three_body_decay=0.57,
+                    three_body_weight=13.4, 
+                    periodic = False):
+  
+
```
Comment (Ryan-Rhys) [inline scripts/gen_FCHL19_rep.py]: Extra blank lines?

File: scripts/gen_FCHL19_rep.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    
+    parser.add_argument('--periodic', type=str2bool, nargs='?', const=True, default=False,
+                        help='Is the system periodic (True/False)?')
+
+    if len(sys.argv) == 1:
+        parser.print_help(sys.stderr)
+        sys.exit(1)
+    args = parser.parse_args()
+    print( args)
+    main(**vars(args))
```
Comment (Ryan-Rhys) [inline scripts/gen_FCHL19_rep.py]: Extra line at end of file. Pylint should catch these.

File: scripts/gen_FCHL19_rep.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-fxyz', type=str, required=True, help='Location of xyz file')
+    parser.add_argument('-fdict', type=str, default='none', help='Location of xyz file '
+                                                                 'that is used for a dictionary')
+    parser.add_argument('--prefix', type=str, default='ASAP', help='Filename prefix')
+    parser.add_argument('--output', type=str, default='xyz', help='The format for output files ([xyz], [matrix])')
+    parser.add_argument('--peratom', type=str2bool, nargs='?', const=True, default=False,
+                        help='Do you want to output per atom representations for multiple frames (True/False)?')
+                        
+    parser.add_argument('--nRs2', type=int, default=24, help='number radial bins in the two-body term')
```
Comment (Ryan-Rhys) [inline scripts/gen_FCHL19_rep.py]: Could we add these descriptions for the parameters to the function docstrings above?

File: scripts/gen_FCHL19_rep.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    Generate the FCHL19 representation.
+
+    Parameters
+    ----------
+    fxyz: string giving location of xyz file
+    fdictxyz: string giving location of xyz file that is used as a dictionary
+    prefix: string giving the filename prefix
+    output: [xyz]: append the FCHL19 representation to extended xyz file; [mat] output as a standlone matrix
+    """
+
```
Comment (Ryan-Rhys) [inline scripts/gen_FCHL19_rep.py]: Extra blank lines?

File: scripts/gen_FCHL19_rep.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    foutput = prefix + "_FCHL19" 
+    desc_name = "FCHL19" 
+
+    # prepare for the output
+    if os.path.isfile(foutput + ".xyz"): os.rename(foutput + ".xyz", "bck." + foutput + ".xyz")
+    if os.path.isfile(foutput + ".desc"): os.rename(foutput + ".desc", "bck." + foutput + ".desc")
+
+    for i, frame in enumerate(frames):
+        #print(i,frame)
```
Comment (Ryan-Rhys) [inline scripts/gen_FCHL19_rep.py]: Do we want to remove the commented out print statements?

Comment (Ryan-Rhys) [reply scripts/gen_FCHL19_rep.py]: Or enforce them?

**Pull request merged**